src/org/python/core/PyReflectedConstructor.java

```diff
@@ -152,26 +152,42 @@ public PyObject __call__(PyObject self, PyObject[] args, String[] keywords) {
                 // Remove the keyword args
                 args = new PyObject[allArgs.length - nkeywords];
                 System.arraycopy(allArgs, 0, args, 0, args.length);
+                Object varargMatch = null;
 
                 // Look for a constructor with no keyword args
                 for (int i = 0; i < n; i++) {
                     rargs = argslist[i];
                     if (rargs.matches(null, args, Py.NoKeywords, callData)) {
-                        method = rargs.method;
-                        break;
+                        if (!argslist[i].isVarArgs) {
+                            method = argslist[i].method;
+                            break;
+                        } else if (varargMatch == null) {
+                            varargMatch = argslist[i].method;
+                        }
                     }
                 }
+                if (method == null) {
+                    method = varargMatch;
+                }
             }
        } else {
            // Just look for a constructor with no keyword args
+            Object varargMatch = null;
            int n = nargs;
            for (int i = 0; i < n; i++) {
                rargs = argslist[i];
                if (rargs.matches(null, args, Py.NoKeywords, callData)) {
-                   method = rargs.method;
-                   break;
+                    if (!argslist[i].isVarArgs) {
+                        method = argslist[i].method;
+                        break;
+                    } else if (varargMatch == null) {
+                        varargMatch = argslist[i].method;
```

Comment:
This still isn't correct.  The previous implementation grabs the first method that matches.  This one simply grabs the last one.  The real problem is that ReflectedArgs.matches() only has visibility into a single declared method's parameter definition and the args passed.  It cannot differentiate between multiple overloaded methods that match to determine which one is best.

The reason that removing #201 works for my overloaded constructors use case is that it does not call `ensureBoxedVarargs()` (because `isVarArgs` is `false`) so it is able to find the correct constructor.

Reply:
It may be that the `varargMethod == null` check is causing problems. If it always chose the last matching vararg method it would pick the correct one for your example are there alternatives where an earlier one matches better than a later one?

Which would java pick out of

```java
Foo(int i, int... rest) {}
Foo(object i, object j, int... rest) {}
```
When called with `Foo(1, 2, 3)`?

I still don't think this is caused by #201. The same thing can be seen in non-constructors CH weren't affected by the change. The reason reverting it fixes this case is that vararg constructors couldn't be called before (without manually passing lists) so they were excluded from the method checks.

Reply:
In Java, it will call `Foo(Object i, Object j, Object k)`...

Reply:
@tpoliaw Here is a tweak to your change that I have confirmed resolves the problem for constructors.   The "best match score" is really just the length of the varargs method parameters list so this accomplishes finding the best match of the varargs constructors if there is no exact match.  

I suspect that the `PyReflectedFunction.__call__()` matching logic needs a similar change:

```python
        } else {
            // Just look for a constructor with no keyword args
            List<ReflectedArgs> varargMatches = new ArrayList<>();
            int n = nargs;
            for (int i = 0; i < n; i++) {
                rargs = argslist[i];
                if (rargs.matches(null, args, Py.NoKeywords, callData)) {
                    if (!argslist[i].isVarArgs) {
                        method = argslist[i].method;
                        break;
                    } else {
                        varargMatches.add(argslist[i]);
                    }
                }
            }
            if (method == null) {
                int argsCount = -1;
                for (ReflectedArgs varargsMatch : varargMatches) {
                    if (varargsMatch.args.length > argsCount) {
                        argsCount = varargsMatch.args.length;
                        method = varargsMatch.method;
                    }
                }
            }
        }
```

Reply:
And if only the first two options are present it's a compile error so Jython will never have to deal with it.

I think removing the `varargMethod == null` check achieves the same thing. If there is  no matching non-vararg method, the longest (and therefore most specific) vararg method will be used instead.

Reply:
> And if only the first two options are present it's a compile error so Jython will never have to deal with it.
> 
> I think removing the `varargMethod == null` check achieves the same thing. If there is no matching non-vararg method, the longest (and therefore most specific) vararg method will be used instead.

I don’t see how removing the `varargMethod == null` check does the same thing UNLESS the ordering is such that the matching varargs constructor with the longest parameter list is ALWAYS the last one encountered.  Is that a safe assumption to make?

Reply:
@tpoliaw Of course, you could collapse the logic to check the length of the ReflectedArgs.args to find the longest matching varargs constructor into the first for loop instead of storing references in a List and then iterating again…

Reply:
>What handles the sorting to ensure that behavior

It's in `ReflectedArgs.compareTo`, there are other checks in the method for when the args length matches but I think this should ensure the sort order.

```java
        int n = this.args.length;
        if (n < oargs.length) {
            return -1;
        }
        if (n > oargs.length) {
            return +1;
        }
```